File: src/main/java/net/bbelovic/adventofcode/day5/Day5PuzzlePart2.java
package net.bbelovic.adventofcode.day5;

public class Day5PuzzlePart2 {
    private static final Rule1 RULE_1 = new Rule1();
    private static final Rule2 RULE_2 = new Rule2();

    public boolean solve(String input) {
        return RULE_1.and(RULE_2).test(input);
    }
}

File: aoc-2020/src/main/kotlin/net/bbelovic/adventofcode/year2020/day5/Day5PuzzlePart2.kt
package net.bbelovic.adventofcode.year2020.day5

class Day5PuzzlePart2 : AbstractDay5Puzzle() {
    override fun solve(input: List<String>): Long {
        val seatIds = input.asSequence()
            .map { each -> computeRowAndCols(each) }
            .toCollection(mutableListOf())
        val max = seatIds.maxOrNull() ?: -1
        val min = seatIds.minOrNull() ?: -1
        var result = 0L
        for (x in min..max) {
            if (!seatIds.contains(x)) {
                result = x
                break
            }
        }
        return result
    }

}
